### PR TITLE
Add constraints on the minimum layout filling fraction

### PR DESCRIPTION
**Pull request by HGSilveri**
- [x]  Adds `BaseDevice.min_layout_filling` to define a minimum layout filling fraction
- [x] Updates the abstract repr schema to support this new parameter

### CODE REVIEW

File: pulser-core/pulser/devices/_device_datacls.py [hunk 1]
```diff
@@ -91,6 +94,9 @@ class BaseDevice(ABC):
         min_layout_traps: The minimum number of traps a layout can have.
         max_layout_traps: An optional value for the maximum number of traps a
             layout can have.
+        min_layout_filling: The smallest fraction of a layout that must be
+            filled with atoms. Only enforced for layouts with more traps than
+            'min_layout_traps'.
```
Comment (pimvenderbosch) [inline pulser-core/pulser/devices/_device_datacls.py:99]: What does the last sentence mean. It shouldn't be possible to have layouts with less traps then `'min_layout_traps'`?

File: pulser-core/pulser/register/register.py [hunk 2]
```diff
@@ -362,14 +362,30 @@ def with_automatic_layout(
...
+        max_traps = device.max_layout_traps
+        if device.min_layout_filling > 0.0:
+            # This is akin to imposing a max number of traps
+            max_allowed_traps = int(
+                len(self.qubit_ids) / device.min_layout_filling
+            )
+            if max_allowed_traps > device.min_layout_traps:
+                # We only enforce min_layout_filling if the maximum number
+                # of traps it allows is greater than the minimum number of
+                # traps required
```
Comment (pimvenderbosch) [inline pulser-core/pulser/register/register.py:374]: Why? Shouldn't this fail otherwise?

File: pulser-core/pulser/register/register.py [hunk 3]
```diff
@@ -362,14 +362,30 @@ def with_automatic_layout(
                 "registers with differentiable coordinates."
             )
 
+        max_traps = device.max_layout_traps
+        if device.min_layout_filling > 0.0:
+            # This is akin to imposing a max number of traps
```
Comment (pimvenderbosch) [inline pulser-core/pulser/register/register.py:367]:
```suggestion
            # This is akin to imposing a max number of traps for a given minimum filling and number of qubits
```

File: pulser-core/pulser/devices/_device_datacls.py (hunk 1 above)
Comment (a-corni) [reply pulser-core/pulser/devices/_device_datacls.py:99]: I agree with you, I find the last sentence a bit weird. 
```suggestion
            filled with atoms. Checked for layouts validating the condition of having more traps than 'min_layout_traps'.
```

Comment (HGSilveri) [reply pulser-core/pulser/devices/_device_datacls.py:99]: Indeed, but they can have exactly `min_layout_traps`, in which case we don't enforce `min_layout_filling` to allow for arbitrarily small registers

Comment (a-corni) [reply pulser-core/pulser/devices/_device_datacls.py:99]: Ah, I didn't have this edge case by reading the sentence...
```suggestion
            filled with atoms. Only enforced for layouts with strictly more traps than
            'min_layout_traps'.
```

File: tests/test_devices.py [hunk 4]
```diff
@@ -564,6 +580,52 @@ def test_layout_filling(helpers, register):
...
+
+    # When the layout has more traps than 'min_layout_traps',
+    # `min_layout_filling` is enforced
+    register = TriangularLatticeLayout(min_traps + 1, 5).hexagonal_register(1)
+    assert register.layout.number_of_traps > min_traps
+    with pytest.raises(MinQubitNumberError):
+        device.validate_layout_filling(register)
+
+    # When the layout has only 'min_layout_traps',
+    # `min_layout_filling` not is enforced
```
[the remark this answers is not in the document]
Comment (HGSilveri) [reply tests/test_devices.py:623]: Hahaha nice
